### PR TITLE
Migrate auth flow to FastAPI; Shutdown Flask App

### PR DESCRIPTION
**Pull request by lvreynoso**
### Summary:
- **What:** Migrates auth flow to FastAPI. Removes last production flask code. Stops flask running on backend container.
- **Ticket:** [sc194360](https://app.shortcut.com/genepi/story/194360)
- **Env:** [https://fastauth-frontend.dev.czgenepi.org/](https://fastauth-frontend.dev.czgenepi.org/)

### Notes:
Depends on #1069 

Upgrades authlib to `1.0.1`
Requires `make local-sync`

### Checklist:
- [x] I merged latest `<base branch>`
- [x] I manually verified the change
- [x] I added labels to my PR
- [x] I tested in multiple browsers
- [x] I added relevant unit tests
- [ ] I have notified others of changes they need to make locally (migrations, jobs, package updates, etc)

### CODE REVIEW

File: src/backend/aspen/api/views/auth.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 from urllib.parse import urlencode
 
 from authlib.integrations.base_client.errors import OAuthError
-from authlib.integrations.starlette_client import StarletteRemoteApp
+from authlib.integrations.starlette_client import StarletteOAuth2App
```
Comment (jgadling) [inline src/backend/aspen/api/views/auth.py:5]: Any particular reason for the upgrade/change?

File: src/backend/aspen/api/views/auth.py [hunk 2]
```diff
@@ -20,21 +20,23 @@
...
+    request: Request,
+    auth0: StarletteOAuth2App = Depends(get_auth0_client),
+    settings: Settings = Depends(get_settings),
 ) -> Response:
     try:
         token = await auth0.authorize_access_token(request)
     except OAuthError:
         raise ex.UnauthorizedException("Invalid token")
-    userinfo = token.get("userinfo")
+    userinfo = token["userinfo"]
```
Comment (jgadling) [inline src/backend/aspen/api/views/auth.py]: Why was this changed? We'll return a 500 error if our access token is missing user info here.

File: src/backend/etc/nginx.conf [hunk 3]
```diff
@@ -41,16 +36,6 @@ http {
...
-      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
-      proxy_set_header X-Forwarded-Proto $scheme;
-      proxy_set_header Host $http_host;
-      # we don't want nginx trying to do something clever with
-      # redirects, we set the Host: header above already.
-      proxy_redirect off;
-      proxy_pass http://flask_server;
-    }
-
     location /v2/ {
```
Comment (jgadling) [inline src/backend/etc/nginx.conf]: Hrm, any issues with updating this to be `location /` and having fastapi handle all requests?

File: src/backend/etc/supervisord.conf [hunk 4]
```diff
@@ -31,18 +31,6 @@ autostart = true
 autorestart = true
 startretries = 9999
 
-[program:flask]
```
Comment (jgadling) [inline src/backend/etc/supervisord.conf:34]: 😍

File: src/frontend/src/common/api/index.tsx [hunk 5]
```diff
@@ -4,8 +4,8 @@ import { jsonToType } from "src/common/utils";
 export enum API {
   USERDATA = "/v2/users/me",
   SAMPLES = "/v2/samples/",
-  LOG_IN = "/login",
-  LOG_OUT = "/logout",
```
Comment (jgadling) [inline src/frontend/src/common/api/index.tsx:8]: 👍

File: src/backend/etc/nginx.conf (hunk 3 above)
Comment (lvreynoso) [reply src/backend/etc/nginx.conf]: Will try that

File: src/backend/aspen/api/views/auth.py (hunk 1 above)
Comment (lvreynoso) [reply src/backend/aspen/api/views/auth.py:5]: I was trying to debug the `userinfo = token.get("userinfo")` line and turns out that syntax only works in newer versions of authlib 1.0.1 not our 0.15.4 version. I could make it work with 0.15.4 but I didn't seem like any more work to change the version number in our poetry file if everything else worked, which it did. So I thought, why not?

Comment (lvreynoso) [reply src/backend/aspen/api/views/auth.py:5]: Oh as for this particular change, it's because the `StarletteRemoteApp` class is gone in 1.0.1.

File: src/backend/aspen/api/views/auth.py (hunk 2 above)
Comment (lvreynoso) [reply src/backend/aspen/api/views/auth.py]: Isn't that what we want? Wouldn't a user just get silently redirected to the splash page otherwise?

File: src/backend/aspen/api/views/auth.py (hunk 1 above)
Comment (jgadling) [reply src/backend/aspen/api/views/auth.py:5]: 👍

File: src/backend/aspen/api/views/auth.py (hunk 2 above)
Comment (jgadling) [reply src/backend/aspen/api/views/auth.py]: ehhh, I'd like to at least control the exception if possible. Return an UnauthorizedException then?

Comment (lvreynoso) [reply src/backend/aspen/api/views/auth.py]: Done!

**Comment by jgadling** [comment]
This isn't working for me at all in local dev -- is there some trick to it?

When I go to the login endpoint I get this in the logs:
```
    return await auth0.authorize_redirect(request, settings.AUTH0_CALLBACK_URL)
  File "/usr/local/lib/python3.9/site-packages/authlib/integrations/starlette_client/apps.py", line 29, in authorize_redirect
    rv = await self.create_authorization_url(redirect_uri, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/authlib/integrations/base_client/async_app.py", line 95, in create_authorization_url
    metadata = await self.load_server_metadata()
  File "/usr/local/lib/python3.9/site-packages/authlib/integrations/base_client/async_app.py", line 76, in load_server_metadata
    resp = await client.request('GET', self._server_metadata_url, withhold_token=True)
  File "/usr/local/lib/python3.9/site-packages/authlib/integrations/httpx_client/oauth2_client.py", line 91, in request
    return await super(AsyncOAuth2Client, self).request(
  File "/usr/local/lib/python3.9/site-packages/httpx/_client.py", line 1494, in request
    response = await self.send(
  File "/usr/local/lib/python3.9/site-packages/httpx/_client.py", line 1586, in send
    response = await self._send_handling_auth(
  File "/usr/local/lib/python3.9/site-packages/httpx/_client.py", line 1616, in _send_handling_auth
    response = await self._send_handling_redirects(
  File "/usr/local/lib/python3.9/site-packages/httpx/_client.py", line 1655, in _send_handling_redirects
    response = await self._send_single_request(request, timeout)
  File "/usr/local/lib/python3.9/site-packages/httpx/_client.py", line 1699, in _send_single_request
    ) = await transport.handle_async_request(
  File "/usr/local/lib/python3.9/site-packages/httpx/_transports/default.py", line 276, in handle_async_request
    (
  File "/usr/local/lib/python3.9/contextlib.py", line 135, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/local/lib/python3.9/site-packages/httpx/_transports/default.py", line 78, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.ConnectError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1123)
```

**Comment by lvreynoso** [comment]
That's weird, I also got that error once but it went away when I reran the `setup_dev_data.sh` script after fixing the `genepi-config` secret string. I don't know if there's any trick to it

**Review by jgadling — APPROVED** [review]
I'm not sure we've 100% tracked down the cause of the cert validation errors in local dev, but if we can get another team member to switch over to this branch without issues, I think we can merge this now and figure out whether we still have a problem later.
